File: src/main/java/com/lcaohoanq/nocket/domain/friendship/Friendship.java
package com.lcaohoanq.nocket.domain.friendship;

import com.lcaohoanq.nocket.base.entity.BaseEntity;
import com.lcaohoanq.nocket.domain.user.User;
import com.lcaohoanq.nocket.enums.FriendShipStatus;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.Table;
import jakarta.persistence.UniqueConstraint;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import lombok.experimental.SuperBuilder;

@Entity
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@SuperBuilder
@Table(name = "friendships",
    uniqueConstraints = {
        @UniqueConstraint(columnNames = {"user1_id", "user2_id"})
    }
)
public class Friendship extends BaseEntity {
    @ManyToOne
    @JoinColumn(name = "user1_id", nullable = false)
    private User user1;

    @ManyToOne
    @JoinColumn(name = "user2_id", nullable = false)
    private User user2;

    @Enumerated(EnumType.STRING)
    @Column(name = "status", nullable = false)
    private FriendShipStatus status;

    // Ensure user1 and user2 are always in a consistent order
    public void normalizeUsers() {
        if (user1.getId().compareTo(user2.getId()) > 0) {
            User temp = user1;
            user1 = user2;
            user2 = temp;
        }
    }
}

File: src/main/java/com/lcaohoanq/nocket/domain/friendship/Friendship.kt
package com.lcaohoanq.nocket.domain.friendship

import com.lcaohoanq.nocket.base.entity.BaseEntity
import com.lcaohoanq.nocket.domain.user.User
import com.lcaohoanq.nocket.enums.FriendShipStatus
import jakarta.persistence.*
import org.hibernate.envers.Audited

@Audited
@Entity
@Table(
    name = "friendships",
    uniqueConstraints = [UniqueConstraint(columnNames = ["user1_id", "user2_id"])]
)
class Friendship : BaseEntity() {
    @ManyToOne
    @JoinColumn(name = "user1_id", nullable = false)
    var user1: User? = null

    @ManyToOne
    @JoinColumn(name = "user2_id", nullable = false)
    var user2: User? = null

    @Enumerated(EnumType.STRING)
    @Column(name = "status", nullable = false)
    var status: FriendShipStatus? = null

    // Ensure user1 and user2 are always in a consistent order
    fun normalizeUsers() {
        if (user1!!.id.compareTo(user2!!.id) > 0) {
            val temp = user1
            user1 = user2
            user2 = temp
        }
    }
}
